### DISCUSSION TITLE
RS-232 diag

### DISCUSSION THREAD

**Discussion by Goshik92** [opencomputeproject/ocp-diag-core#649] [Ideas]
I received a request to implement an RS-232 diag to test if the adapter functions correctly when it is not connected to the target device. The OCP TV initiative members may also find such diag useful, so I am proposing to implement it as an OCP diag. 

The following functionality may be included in the diag:
- Presence check. This test will check if `/dev/ttyUSB0` (the name will not be hardcoded) is present in the system to ensure that the driver is able to initialize the adapter.
- Hardware loopback check. The user will insert a loopback plug in the adapter's port and run the test that will send a significant amount of data through the adapter to ensure that the data is mirrored correctly.
- Software loopback check. This test will enable the software loopback mode (the method may vary from adapter to adapter, so this test is not universal) and perform the same procedure as in the hardware loopback check. Even though it is unclear how much value this test provides in addition to the presence check, the customer finds it useful as it is perceived to create traffic on the PCI-e bus (in case of a PCI-E RS-232 adapter).

**Comment by mimir-d** [comment]
- i'd say go for it. My suggestion is though, if you want it to be easy to use by the community, start by having github the source of truth for the codebase (as opposed to using tooling to export code from internal repos).
- If you'd like this to be an ocptv diag, I can make a repo for it as long as we have a clear maintenance model (ie. are you signing up for this?).
- for the software loopback, i remember you mentioned this is specific to the device and/or driver. Ideally we either use a driver which is universally available (eg in kernel or out of tree OSS)

**Reply by Goshik92** [reply]
1. Agree. That was my original plan.
2. Can remain a maintainer as long as I am involved in OCP.
3. I think my next comment will address this.

**Comment by Goshik92** [comment]
There is an interesting open-source MIT license utility: [linux-serial-test](https://github.com/cbrake). It seems to have a lot of useful features, including support for software loopback (it does not mention any specific hardware, so apparently there is a more or less universal mechanism to enable it):
```
Usage: linux-serial-test [OPTION]

  -h, --help
  -b, --baud               Baud rate, 115200, etc (115200 is default)
  -p, --port               Port (/dev/ttyS0, etc) (must be specified)
  -d, --divisor            UART Baud rate divisor (can be used to set custom baud rates)
  -R, --rx_dump            Dump Rx data (ascii, raw)
  -T, --detailed_tx        Detailed Tx data
  -s, --stats              Dump serial port stats every 5s
  -S, --stop-on-err        Stop program if we encounter an error
  -y, --single-byte        Send specified byte to the serial port
  -z, --second-byte        Send another specified byte to the serial port
  -c, --rts-cts            Enable RTS/CTS flow control
  -B, --2-stop-bit         Use two stop bits per character
  -P, --parity             Use parity bit (odd, even, mark, space)
  -k, --loopback           Use internal hardware loop back
  -K, --write-follow       Write follows the read count (can be used for multi-serial loopback)
  -e, --dump-err           Display errors
  -r, --no-rx              Don't receive data (can be used to test flow control)
                           when serial driver buffer is full
  -t, --no-tx              Don't transmit data
  -l, --rx-delay           Delay between reading data (ms) (can be used to test flow control)
  -a, --tx-delay           Delay between writing data (ms)
  -w, --tx-bytes           Number of bytes for each write (default is to repeatedly write 1024 bytes
                           until no more are accepted)
  -q, --rs485              Enable RS485 direction control on port, and set delay from when TX is
                           finished and RS485 driver enable is de-asserted. Delay is specified in
                           bit times. To optionally specify a delay from when the driver is enabled
                           to start of TX use 'after_delay.before_delay' (-q 1.1)
  -Q, --rs485_rts          Deassert RTS on send, assert after send. Omitting -Q inverts this logic.
  -o, --tx-time            Number of seconds to transmit for (defaults to 0, meaning no limit)
  -i, --rx-time            Number of seconds to receive for (defaults to 0, meaning no limit)
  -A, --ascii              Output bytes range from 32 to 126 (default is 0 to 255)
  -I, --rx-timeout         Receive timeout
  -O, --tx-timeout         Transmission timeout
  -W, --tx-wait            Number of seconds to wait before to transmit (defaults to 0, meaning no wait)
  -Z, --error-on-timeout   Treat timeouts as errors
  -n, --no-icount          Do not request driver for counts of input serial line interrupts (TIOCGICOUNT)
  -f, --flush-buffers      Flush RX and TX buffers before starting
```

I am considering wrapping it into an OCP diag. There are multiple ways of doing it:
- Call the utility from a Python script and parse its output (same approach as in Memtester). Could be ideal if the utility was in distros (could not find it), otherwise the user would need to clone and compile it to run the diag, which is inconvenient.
- Fork the utility and replace `printf`s with corresponding OCP TV calls. This will probably require using `ocp-diag-core-cpp` and `bazel`.

**Comment by Goshik92** [comment]
Due to an internal decision, this diag will be based on Google's internal testing framework. OCP TV members are free to take over the development of its open-source version if necessary.
